source/classes/Westley.js

```diff
@@ -58,24 +41,42 @@
 			throw new Error("Invalid command");
 		}
 		var commandComponents = commandTools.extractCommandComponents(command),
-			commandKey = commandComponents.shift();//,
-			//commandFilename = "command." + commandKey + ".js",
-			//commandExecute;
-		if (!availableCommands.hasOwnProperty(commandKey)) {
-			throw new Error("Unrecognised command: " + commandKey);
-		}
-		// try {
-		// 	commandExecute = require("__buttercup/classes/commands/" + commandFilename);
-		// } catch (err) {
-		// 	throw new Error("Unrecognised command: " + commandKey);
-		// }
-		var commandToExecute = availableCommands[commandKey];
+			commandKey = commandComponents.shift();
+
+		var commandObject = this._getCommandForKey(commandKey);
 		this._history.push(command);
-		commandToExecute.apply(commandToExecute, [this._dataset].concat(commandComponents));
+		commandObject.execute.apply(commandObject, [this._dataset].concat(commandComponents));
 		return this;
 	};
 
 	/**
+	 * Gets a command by its key from the cache with its dependencies injected
+	 * @param {String} commandKey The key of the command
+	 * @returns {Command} Returns the command
+	 * @memberof Westley
+	 */
+	Westley.prototype._getCommandForKey = function(commandKey) {
+		// If the command doesn't exist in the cache
+		if (this._cachedCommands[commandKey] === undefined) {
+			// Get the command object and inject its dependencies
+			var requirement = new (require("__buttercup/classes/commands/command." + commandKey + ".js"))();
+
+			if (requirement.injectSearching !== undefined) {
+				requirement.injectSearching(searching);
+			}
+
+			if (requirement.injectEntry !== undefined) {
+				requirement.injectEntry(entry);
+			}
```

Comment:
By having the commands `require` their own dependencies, we have no control over how they get them. That means there is no way to fake those in order to have a unit test for the behaviour of the commands without also testing the dependencies at the same time.

Take a look online at Dependency Injection and Inversion of Control, it really is a useful thing when you get the hang of it :smile: